### PR TITLE
Use variable template versions of std traits

### PR DESCRIPTION
**Pull request by bernhardmgruber**
This PR refactores the code to use the variable template versions of standard traits.
It furthermore contains a refactoring of `core/Cuda.hpp` and `core/Hip.hpp` into a common `core/CudaHipCommon.hpp` header, triggered by a bad clang-format formatting.

- [x] Please merge first: #1593

### CODE REVIEW

File: include/alpaka/core/Cuda.hpp [hunk 1]
```diff
@@ -101,37 +101,16 @@ namespace alpaka
...
... std::is_same_v<T, uint1> || std::is_same_v<T, ulong1> || std::is_same_v<T, ulonglong1> || std::is_same_v<T, ushort1> || std::is_same_v<T, char2> || std::is_same_v<T, double2> || std::is_same_v<T, float2> || std::is_same_v<T, int2> || std::is_same_v<T, long2> || std::is_same_v<T, longlong2> || std::is_same_v<T, short2> || std::is_same_v<T, uchar2> || std::is_same_v<T, uint2> || std::is_same_v<T, ulong2> || std::is_same_v<T, ulonglong2> || std::is_same_v<T, ushort2> || std::is_same_v<T, char3> || std::is_same_v<T, dim3> || std::is_same_v<T, double3> || std::is_same_v<T, float3> || std::is_same_v<T, int3> || std::is_same_v<T, long3> || std::is_same_v<T, longlong3> || std::is_same_v<T, short3> || std::is_same_v<T, uchar3> || std::is_same_v<T, uint3> || std::is_same_v<T, ulong3> || std::is_same_v<T, ulonglong3> || std::is_same_v<T, ushort3> || std::is_same_v<T, char4> || std::is_same_v<T, double4> || std::is_same_v<T, float4> || std::is_same_v<T, int4> || std::is_same_v<T, long4> || std::is_same_v<T, longlong4> || std::is_same_v<T, short4> || std::is_same_v<T, uchar4> || std::is_same_v<T, uint4> || std::is_same_v<T, ulong4> || std::is_same_v<T, ulonglong4> || std::is_same_v<T, ushort4>
```
Comment (SimeonEhrig) [inline include/alpaka/core/Cuda.hpp]: Looks like, that clang-format failed :thinking: Also on various other places.

Comment (bernhardmgruber) [reply include/alpaka/core/Cuda.hpp]: This is indeed a major mess. I will see what I can do here.

Alternatively, this line could also just become: `boost::mp11::mp_contains<boost::mp11::mp_list<char1, double1, float1, ...>, T>::value`.

Comment (bernhardmgruber) [reply include/alpaka/core/Cuda.hpp]: Fixed using a type list. Fortunately, we had the necessary meta functions in alpaka.

File: test/unit/atomic/src/AtomicTest.cpp [hunk 2]
```diff
@@ -581,9 +575,9 @@ class AtomicTestKernel<
     alpaka::AccGpuCudaRt<TDim, TIdx>,
     T,
     std::enable_if_t<
-        !std::is_same<int, T>::value && !std::is_same<unsigned int, T>::value
-        && !std::is_same<unsigned long int, T>::value && !std::is_same<unsigned long long int, T>::value
-        && !std::is_same<float, T>::value && !std::is_same<double, T>::value>>
+        !std::is_same_v<
+            int,
+            T> && !std::is_same_v<unsigned int, T> && !std::is_same_v<unsigned long int, T> && !std::is_same_v<unsigned long long int, T> && !std::is_same_v<float, T> && !std::is_same_v<double, T>>>
```
Comment (j-stephan) [inline test/unit/atomic/src/AtomicTest.cpp]: clang-format seems to fail again.

File: test/unit/atomic/src/AtomicTest.cpp [hunk 3]
```diff
@@ -801,9 +795,9 @@ class AtomicTestKernel<
     alpaka::AccGpuHipRt<TDim, TIdx>,
     T,
     std::enable_if_t<
-        !std::is_same<int, T>::value && !std::is_same<unsigned int, T>::value
-        && !std::is_same<unsigned long int, T>::value && !std::is_same<unsigned long long int, T>::value
-        && !std::is_same<float, T>::value && !std::is_same<double, T>::value>>
+        !std::is_same_v<
+            int,
+            T> && !std::is_same_v<unsigned int, T> && !std::is_same_v<unsigned long int, T> && !std::is_same_v<unsigned long long int, T> && !std::is_same_v<float, T> && !std::is_same_v<double, T>>>
```
Comment (j-stephan) [inline test/unit/atomic/src/AtomicTest.cpp]: And here.

**Review by j-stephan — CHANGES_REQUESTED** [review]
See the formatting issues above.

**Review by j-stephan — APPROVED** [review]

**Pull request merged**